File: sdks/python/memsy/resources/roles.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from memsy.models import RoleResource

if TYPE_CHECKING:
    from memsy.async_client import AsyncMemsyClient
    from memsy.client import MemsyClient


class RolesResource:
    """Sync wrapper for memsy-core /roles onboarding endpoints."""

    def __init__(self, client: MemsyClient) -> None:
        self._client = client

    def list(self, org_id: str, *, limit: int = 100, offset: int = 0) -> list[RoleResource]:
        """List role customization records for an org."""
        params = {"org_id": org_id, "limit": limit, "offset": offset}
        data, _, _ = self._client._request("GET", "/roles", params=params)
        return [RoleResource.from_dict(r) for r in (data or [])]

    def create(self, org_id: str, name: str, focus: str) -> RoleResource:
        """
        Create a role customization record.

        :param org_id: The org this role belongs to.
        :param name: Display name for the role.
        :param focus: Description of the role's memory promotion scope.
        """
        body = {"org_id": org_id, "name": name, "focus": focus}
        data, _, _ = self._client._request("POST", "/roles", json=body)
        return RoleResource.from_dict(data)

    def get(self, role_id: str, org_id: str) -> RoleResource:
        """Retrieve a single role customization record."""
        data, _, _ = self._client._request("GET", f"/roles/{role_id}", params={"org_id": org_id})
        return RoleResource.from_dict(data)

    def update(
        self,
        role_id: str,
        org_id: str,
        *,
        name: str | None = None,
        focus: str | None = None,
        promotion_prompt: str | None = None,
    ) -> RoleResource:
        """Partially update a role customization record."""
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if focus is not None:
            body["focus"] = focus
        if promotion_prompt is not None:
            body["promotion_prompt"] = promotion_prompt
        data, _, _ = self._client._request(
            "PATCH", f"/roles/{role_id}", params={"org_id": org_id}, json=body
        )
        return RoleResource.from_dict(data)

    def regenerate_prompt(self, role_id: str, org_id: str) -> RoleResource:
        """Re-run the LLM to regenerate the promotion_prompt for this role."""
        data, _, _ = self._client._request(
            "POST", f"/roles/{role_id}/regenerate-prompt", params={"org_id": org_id}
        )
        return RoleResource.from_dict(data)

    def delete(self, role_id: str, org_id: str) -> None:
        """Remove the role customization record. Memories with this role_id are unaffected."""
        self._client._request("DELETE", f"/roles/{role_id}", params={"org_id": org_id})


class AsyncRolesResource:
    """Async wrapper for memsy-core /roles onboarding endpoints."""

    def __init__(self, client: AsyncMemsyClient) -> None:
        self._client = client

    async def list(self, org_id: str, *, limit: int = 100, offset: int = 0) -> list[RoleResource]:
        """List role customization records for an org."""
        params = {"org_id": org_id, "limit": limit, "offset": offset}
        data, _, _ = await self._client._request("GET", "/roles", params=params)
        return [RoleResource.from_dict(r) for r in (data or [])]

    async def create(self, org_id: str, name: str, focus: str) -> RoleResource:
        """
        Create a role customization record.

        :param org_id: The org this role belongs to.
        :param name: Display name for the role.
        :param focus: Description of the role's memory promotion scope.
        """
        body = {"org_id": org_id, "name": name, "focus": focus}
        data, _, _ = await self._client._request("POST", "/roles", json=body)
        return RoleResource.from_dict(data)

    async def get(self, role_id: str, org_id: str) -> RoleResource:
        """Retrieve a single role customization record."""
        data, _, _ = await self._client._request(
            "GET", f"/roles/{role_id}", params={"org_id": org_id}
        )
        return RoleResource.from_dict(data)

    async def update(
        self,
        role_id: str,
        org_id: str,
        *,
        name: str | None = None,
        focus: str | None = None,
        promotion_prompt: str | None = None,
    ) -> RoleResource:
        """Partially update a role customization record."""
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if focus is not None:
            body["focus"] = focus
        if promotion_prompt is not None:
            body["promotion_prompt"] = promotion_prompt
        data, _, _ = await self._client._request(
            "PATCH", f"/roles/{role_id}", params={"org_id": org_id}, json=body
        )
        return RoleResource.from_dict(data)

    async def regenerate_prompt(self, role_id: str, org_id: str) -> RoleResource:
        """Re-run the LLM to regenerate the promotion_prompt for this role."""
        data, _, _ = await self._client._request(
            "POST", f"/roles/{role_id}/regenerate-prompt", params={"org_id": org_id}
        )
        return RoleResource.from_dict(data)

    async def delete(self, role_id: str, org_id: str) -> None:
        """Remove the role customization record. Memories with this role_id are unaffected."""
        await self._client._request("DELETE", f"/roles/{role_id}", params={"org_id": org_id})

File: sdks/node/src/resources/roles.ts
import type { BaseHttpClient } from "../http.js";
import {
  type OnboardingUpdate,
  type Role,
  buildOnboardingUpdateBody,
  parseRole,
} from "../models.js";

export interface RoleListOptions {
  limit?: number;
  offset?: number;
}

export type RoleUpdate = OnboardingUpdate;

export class RolesResource {
  constructor(private readonly client: BaseHttpClient) {}

  async list(orgId: string, options: RoleListOptions = {}): Promise<Role[]> {
    const { data } = await this.client.request<Record<string, unknown>[] | null>("GET", "/roles", {
      query: { org_id: orgId, limit: options.limit ?? 100, offset: options.offset ?? 0 },
    });
    return (data ?? []).map(parseRole);
  }

  async create(orgId: string, name: string, focus: string): Promise<Role> {
    const { data } = await this.client.request<Record<string, unknown>>("POST", "/roles", {
      body: { org_id: orgId, name, focus },
    });
    return parseRole(data);
  }

  async get(roleId: string, orgId: string): Promise<Role> {
    const { data } = await this.client.request<Record<string, unknown>>(
      "GET",
      `/roles/${encodeURIComponent(roleId)}`,
      { query: { org_id: orgId } }
    );
    return parseRole(data);
  }

  async update(roleId: string, orgId: string, update: RoleUpdate): Promise<Role> {
    const { data } = await this.client.request<Record<string, unknown>>(
      "PATCH",
      `/roles/${encodeURIComponent(roleId)}`,
      { body: buildOnboardingUpdateBody(update), query: { org_id: orgId } }
    );
    return parseRole(data);
  }

  async regeneratePrompt(roleId: string, orgId: string): Promise<Role> {
    const { data } = await this.client.request<Record<string, unknown>>(
      "POST",
      `/roles/${encodeURIComponent(roleId)}/regenerate-prompt`,
      { query: { org_id: orgId } }
    );
    return parseRole(data);
  }

  async delete(roleId: string, orgId: string): Promise<void> {
    await this.client.request<null>("DELETE", `/roles/${encodeURIComponent(roleId)}`, {
      query: { org_id: orgId },
    });
  }
}
